mlflow/store/artifact/s3_artifact_repo.py

```diff
@@ -97,26 +115,47 @@ def _get_s3_client(
         access_key_id=access_key_id,
         secret_access_key=secret_access_key,
         session_token=session_token,
+        region_name=region_name,
     )
 
 
-class S3ArtifactRepository(ArtifactRepository):
+class S3ArtifactRepository(CloudArtifactRepository):
     """Stores artifacts on Amazon S3."""
 
     def __init__(
-        self, artifact_uri, access_key_id=None, secret_access_key=None, session_token=None
+        self,
+        artifact_uri,
+        access_key_id=None,
+        secret_access_key=None,
+        session_token=None,
+        addressing_style="path",
+        use_optimized=False,
     ):
         super().__init__(artifact_uri)
         self._access_key_id = access_key_id
         self._secret_access_key = secret_access_key
         self._session_token = session_token
+        self._addressing_style = addressing_style
+        self.bucket, self.bucket_path = self.parse_s3_compliant_uri(self.artifact_uri)
+        self._use_optimized = use_optimized
+        self._region_name = self._get_region_name()
+
+    def _get_region_name(self):
+        temp_client = _get_s3_client(
+            addressing_style=self._addressing_style,
+            access_key_id=self._access_key_id,
+            secret_access_key=self._secret_access_key,
+            session_token=self._session_token,
+        )
+        return temp_client.get_bucket_location(Bucket=self.bucket)["LocationConstraint"]
```

Comment:
@dbczumar let me know if this should just return `None` when `self._use_optimized` is `False`. The region name is used when uploading files for presigned URLs.